### PR TITLE
Memory scoping

### PR DESCRIPTION
**Pull request by iamdefinitelyahuman**
### What I did
Deallocate internal memory variables between statements.

### How I did it
Upon exiting `internal_memory_scope`, all internal variables declared within the scope are deallocated. The scope is applied prior to parsing each statement node.

### How to verify it
Run tests.

### Cute Animal Picture
image

### CODE REVIEW

File: vyper/parser/stmt.py [hunk 1]
```diff
@@ -38,7 +38,9 @@ def __init__(self, node: vy_ast.VyperNode, context: Context) -> None:
         if fn is None:
             raise TypeCheckFailure(f"Invalid statement node: {type(node).__name__}")
 
-        self.lll_node = fn()
+        with context.internal_memory_scope():
+            self.lll_node = fn()
```
Comment (fubuloubu) [inline vyper/parser/stmt.py:42]: nice

**Review by fubuloubu — COMMENTED** [review]
Should there be some tests in here?

File: vyper/parser/function_definitions/parse_function.py [hunk 2]
```diff
@@ -1,11 +1,11 @@
+from vyper.parser import memory_allocator as mem_alloc
```
Comment (fubuloubu) [inline vyper/parser/function_definitions/parse_function.py:1]: why not import the class?

**Review by fubuloubu — APPROVED** [review]

File: vyper/parser/function_definitions/parse_function.py (hunk 2 above)
Comment (iamdefinitelyahuman) [reply vyper/parser/function_definitions/parse_function.py:1]: `from [module] import [object]` doesn't work with mocking :(

Comment (iamdefinitelyahuman) [reply vyper/parser/function_definitions/parse_function.py:2]: It breaks mocking.  I added a comment to explain.

**Comment by fubuloubu** [comment]
fuck you codecov

**Review by fubuloubu — APPROVED** [review]

**Comment by fubuloubu** [comment]
Shut the hell up lgtm

File: vyper/parser/function_definitions/parse_function.py [hunk 3]
```diff
@@ -59,6 +61,6 @@ def parse_function(code, sigs, origcode, global_ctx, is_contract_payable, _vars=
         )
 
     o.context = context
-    o.total_gas = o.gas + calc_mem_gas(o.context.memory_allocator.get_next_memory_position())
+    o.total_gas = o.gas + calc_mem_gas(o.context.memory_allocator.size_of_mem)
```
Comment (charles-cooper) [inline vyper/parser/function_definitions/parse_function.py:64]: this should be part of the allocator API - `allocator.memsize()`.

File: vyper/parser/context.py [hunk 4]
```diff
@@ -104,28 +104,64 @@ def range_scope(self):
...
-        # syntactic sugar for `make_blockscope` used to release
-        # memory after creating temporary internal variables
-        return self.make_blockscope(scope_id)
+    @contextlib.contextmanager
+    def internal_memory_scope(self):
+        """
+        Internal memory scope context manager.
+
+        Internal variables that are declared within this context are de-allocated
+        upon exitting the context.
```
Comment (charles-cooper) [inline vyper/parser/context.py:113]: exiting

File: vyper/parser/context.py [hunk 5]
```diff
@@ -104,28 +104,64 @@ def range_scope(self):
...
             del self.vars[name]
 
         # Remove block scopes
-        self.blockscopes.remove(blockscope_id)
+        self._scopes.remove(scope_id)
 
-    def _new_variable(self, name: str, typ: NodeType, var_pos: int) -> int:
+    def _new_variable(self, name: str, typ: NodeType, var_size: int, is_internal: bool) -> int:
+        if is_internal:
+            var_pos = self.memory_allocator.expand_memory(var_size)
```
Comment (charles-cooper) [inline vyper/parser/context.py:155]: why do internal variables get a different allocation strategy?

File: vyper/parser/context.py [hunk 6]
```diff
@@ -104,28 +104,64 @@ def range_scope(self):
...
+        else:
+            var_pos = self.memory_allocator.allocate_memory(var_size)
         self.vars[name] = VariableRecord(
-            name=name, pos=var_pos, typ=typ, mutable=True, blockscopes=self.blockscopes.copy(),
+            name=name,
+            pos=var_pos,
+            typ=typ,
+            mutable=True,
+            blockscopes=self._scopes.copy(),
```
Comment (charles-cooper) [inline vyper/parser/context.py:163]: i'm not sure why a variable needs to have pointers to all parent scopes. this seems like it should just be a single scope_id.

**Review by charles-cooper — COMMENTED** [review]
this code look really cleaned up compared to before! found a single issue in `partially_allocate` (the `size < self.size` caller invariant might change in the future so it's good hygiene to fix that assertion) and some questions / style suggestions.

File: vyper/parser/memory_allocator.py [hunk 7]
```diff
@@ -7,13 +7,34 @@
...
+        ---------
+        size : int
+            Number of bytes to allocate
+
+        Returns
+        -------
+        int
+            Position of the newly allocated memory
+        """
+        if size > self.size:
```
[the remark this answers is not in the document]
Comment (iamdefinitelyahuman) [reply vyper/parser/memory_allocator.py:31]: Good catch!

File: vyper/parser/memory_allocator.py [hunk 8]
```diff
@@ -101,9 +129,8 @@ def deallocate_memory(self, pos: int, size: int) -> None:
         # releasing from the end of the allocated memory - reduce the free memory pointer
         if pos + size == self.next_mem:
```
[the remark this answers is not in the document]
Comment (iamdefinitelyahuman) [reply vyper/parser/memory_allocator.py:130]: Yeah, all of this logic can be cleaned up quite a bit.

File: vyper/parser/context.py (hunk 5 above)
Comment (iamdefinitelyahuman) [reply vyper/parser/context.py:155]: There are places in `parser` where an assumption is made that internal variables will be allocated sequentially within memory. If we don't do this, we end up with memory corruption.  I tried to unravel / refactor that logic but eventually decided it wasn't safe - I can't be sure I've found them all.

By always allocating new internal variables on the edge of memory like this, we don't risk memory corruption but still gain some efficiency by not expanding memory needlessly.  In the future when we refactor `parser` we should eliminate the assumption about how memory is sequenced, and then we can do away with this logic.

File: vyper/parser/context.py (hunk 6 above)
Comment (iamdefinitelyahuman) [reply vyper/parser/context.py:163]: Good point.  I just copied this from existing code without thinking :grimacing:

File: vyper/parser/context.py (hunk 5 above)
Comment (charles-cooper) [reply vyper/parser/context.py:155]: weird. gotcha

**Pull request merged**